File: src/lib/analytics/caffeine-sleep.ts
import { type ConfidenceLevel, getConfidenceLevel } from './correlation';
import { localMinutesOfDay } from './local-time';
import { welchTTest, mean } from './stats';
import { DEFAULT_CAFFEINE_CUTOFF_HOUR } from './constants.generated';

export type CaffeineSleepResult = {
	/** A personal cutoff, only when the split survives a multiplicity-corrected test. */
	estimatedCutoffHour: number | null;
	/**
	 * Literature default to fall back on: 400 mg six hours before bed still cost
	 * over an hour of sleep (Drake 2013) and the meta-analytic abstention window
	 * is ~9 h (Gardiner 2023), i.e. ~14:00 for a 23:00 bedtime.
	 */
	defaultCutoffHour: number;
	/** Bonferroni-corrected p-value of the best split, null when nothing was testable. */
	pValue: number | null;
	/** Number of candidate cutoffs that had enough nights on both sides to test. */
	comparisons: number;
	hourlyImpact: { hour: number; avgQuality: number; avgDuration: number; count: number }[];
	confidence: ConfidenceLevel;
	sampleSize: number;
};

const MIN_NIGHTS_PER_SIDE = 5;
const MIN_QUALITY_DELTA = 0.5;
const SIGNIFICANCE = 0.05;

function getNextDate(dateStr: string): string {
	const [year, month, day] = dateStr.split('-').map(Number);
	const d = new Date(Date.UTC(year, month - 1, day + 1));
	return d.toISOString().slice(0, 10);
}

/**
 * Buckets nights by the hour of the day's last caffeine and scans cutoffs
 * 12:00–20:00 for the split with the strongest quality drop. Nine candidate
 * split points searched for a maximum is a maximally-selected statistic, so
 * the winner's Welch p-value is Bonferroni-corrected over the candidates that
 * were actually testable before it may replace the literature default.
 */
export function computeCaffeineSleepCutoff(
	caffeineEntries: { date: string; eatenAt: string | null; caffeine: number }[],
	sleepData: { date: string; sleepQuality: number | null; sleepDurationMinutes: number | null }[],
	timeZone: string
): CaffeineSleepResult {
	const sleepByDate = new Map<string, { quality: number; duration: number }>();
	for (const s of sleepData) {
		if (s.sleepQuality !== null && s.sleepDurationMinutes !== null) {
			sleepByDate.set(s.date, { quality: s.sleepQuality, duration: s.sleepDurationMinutes });
		}
	}

	const lastCaffeineHourByDate = new Map<string, number>();
	for (const entry of caffeineEntries) {
		if (!entry.eatenAt || entry.caffeine <= 0) continue;
		const minutes = localMinutesOfDay(entry.eatenAt, timeZone);
		if (minutes === null) continue;
		const hour = Math.floor(minutes / 60);
		const existing = lastCaffeineHourByDate.get(entry.date);
		if (existing === undefined || hour > existing) {
			lastCaffeineHourByDate.set(entry.date, hour);
		}
	}

	const hourBuckets = new Map<number, { quality: number[]; duration: number[] }>();
	for (const [date, lastHour] of lastCaffeineHourByDate) {
		const nextDate = getNextDate(date);
		const sleep = sleepByDate.get(nextDate);
		if (!sleep) continue;

		if (!hourBuckets.has(lastHour)) hourBuckets.set(lastHour, { quality: [], duration: [] });
		hourBuckets.get(lastHour)!.quality.push(sleep.quality);
		hourBuckets.get(lastHour)!.duration.push(sleep.duration);
	}

	const sortedBuckets = [...hourBuckets.entries()].sort(([a], [b]) => a - b);
	const hourlyImpact = sortedBuckets.map(([hour, { quality, duration }]) => ({
		hour,
		avgQuality: mean(quality),
		avgDuration: mean(duration),
		count: quality.length
	}));

	const sampleSize = hourlyImpact.reduce((s, h) => s + h.count, 0);

	let comparisons = 0;
	let bestCandidate: number | null = null;
	let bestP = 1;

	for (let candidate = 12; candidate <= 20; candidate++) {
		const before = sortedBuckets.filter(([hour]) => hour < candidate).flatMap(([, b]) => b.quality);
		const after = sortedBuckets.filter(([hour]) => hour >= candidate).flatMap(([, b]) => b.quality);
		if (before.length < MIN_NIGHTS_PER_SIDE || after.length < MIN_NIGHTS_PER_SIDE) continue;

		comparisons++;
		const delta = mean(before) - mean(after);
		if (delta <= MIN_QUALITY_DELTA) continue;

		const { pValue } = welchTTest(before, after);
		if (pValue < bestP) {
			bestP = pValue;
			bestCandidate = candidate;
		}
	}

	const correctedP = comparisons > 0 ? Math.min(1, bestP * comparisons) : null;
	const estimatedCutoffHour =
		bestCandidate !== null && correctedP !== null && correctedP < SIGNIFICANCE
			? bestCandidate
			: null;

	return {
		estimatedCutoffHour,
		defaultCutoffHour: DEFAULT_CAFFEINE_CUTOFF_HOUR,
		pValue: correctedP,
		comparisons,
		hourlyImpact,
		confidence: getConfidenceLevel(sampleSize),
		sampleSize
	};
}

File: mobile/shared/src/commonMain/kotlin/com/bissbilanz/analytics/CaffeineSleep.kt
package com.bissbilanz.analytics

import kotlin.math.min

data class HourlyImpact(
    val hour: Int,
    val avgQuality: Double,
    val avgDuration: Double,
    val count: Int,
)

data class CaffeineSleepResult(
    /** A personal cutoff, only when the split survives a multiplicity-corrected test. */
    val estimatedCutoffHour: Int?,
    /** Literature default to fall back on (~9 h before a 23:00 bedtime; Drake 2013, Gardiner 2023). */
    val defaultCutoffHour: Int,
    /** Bonferroni-corrected p-value of the best split, null when nothing was testable. */
    val pValue: Double?,
    /** Number of candidate cutoffs that had enough nights on both sides to test. */
    val comparisons: Int,
    val hourlyImpact: List<HourlyImpact>,
    val confidence: ConfidenceLevel,
    val sampleSize: Int,
)

data class CaffeineEntry(
    val date: String,
    val eatenAt: String?,
    val caffeine: Double,
)

data class SleepDataPoint(
    val date: String,
    val sleepQuality: Double?,
    val sleepDurationMinutes: Double?,
)

private const val MIN_NIGHTS_PER_SIDE = 5
private const val MIN_QUALITY_DELTA = 0.5
private const val SIGNIFICANCE = 0.05

/**
 * Buckets nights by the hour of the day's last caffeine and scans cutoffs
 * 12:00–20:00 for the split with the strongest quality drop. Nine candidate
 * split points searched for a maximum is a maximally-selected statistic, so the
 * winner's Welch p-value is Bonferroni-corrected over the candidates that were
 * actually testable before it may replace the literature default.
 */
fun computeCaffeineSleepCutoff(
    caffeineEntries: List<CaffeineEntry>,
    sleepData: List<SleepDataPoint>,
    timeZone: String,
): CaffeineSleepResult {
    val sleepByDate = mutableMapOf<String, Pair<Double, Double>>()
    for (s in sleepData) {
        if (s.sleepQuality != null && s.sleepDurationMinutes != null) {
            sleepByDate[s.date] = Pair(s.sleepQuality, s.sleepDurationMinutes)
        }
    }

    val lastCaffeineHourByDate = mutableMapOf<String, Int>()
    for (entry in caffeineEntries) {
        if (entry.eatenAt == null || entry.caffeine <= 0) continue
        val localMinutes = localMinutesOfDay(entry.eatenAt, timeZone) ?: continue
        val hour = localMinutes / 60
        val existing = lastCaffeineHourByDate[entry.date]
        if (existing == null || hour > existing) lastCaffeineHourByDate[entry.date] = hour
    }

    val hourBuckets = mutableMapOf<Int, Pair<MutableList<Double>, MutableList<Double>>>()
    for ((date, lastHour) in lastCaffeineHourByDate) {
        val nextDate = shiftDate(date, 1)
        val sleep = sleepByDate[nextDate] ?: continue
        val bucket = hourBuckets.getOrPut(lastHour) { Pair(mutableListOf(), mutableListOf()) }
        bucket.first.add(sleep.first)
        bucket.second.add(sleep.second)
    }

    val sortedBuckets = hourBuckets.entries.sortedBy { it.key }
    val hourlyImpact =
        sortedBuckets.map { (hour, pair) ->
            HourlyImpact(
                hour = hour,
                avgQuality = mean(pair.first),
                avgDuration = mean(pair.second),
                count = pair.first.size,
            )
        }

    val sampleSize = hourlyImpact.sumOf { it.count }

    var comparisons = 0
    var bestCandidate: Int? = null
    var bestP = 1.0
    for (candidate in 12..20) {
        val before = sortedBuckets.filter { it.key < candidate }.flatMap { it.value.first }
        val after = sortedBuckets.filter { it.key >= candidate }.flatMap { it.value.first }
        if (before.size < MIN_NIGHTS_PER_SIDE || after.size < MIN_NIGHTS_PER_SIDE) continue
        comparisons++
        val delta = mean(before) - mean(after)
        if (delta <= MIN_QUALITY_DELTA) continue
        val pValue = welchTTest(before, after).pValue
        if (pValue < bestP) {
            bestP = pValue
            bestCandidate = candidate
        }
    }

    val correctedP = if (comparisons > 0) min(1.0, bestP * comparisons) else null
    val estimatedCutoffHour =
        if (bestCandidate != null && correctedP != null && correctedP < SIGNIFICANCE) bestCandidate else null

    return CaffeineSleepResult(
        estimatedCutoffHour = estimatedCutoffHour,
        defaultCutoffHour = DEFAULT_CAFFEINE_CUTOFF_HOUR,
        pValue = correctedP,
        comparisons = comparisons,
        hourlyImpact = hourlyImpact,
        confidence = getConfidenceLevel(sampleSize),
        sampleSize = sampleSize,
    )
}
